### ISSUE TITLE
Interfaces with all members optional is almost the same as 'any'

### ISSUE THREAD

**Issue by bman61** [Microsoft/TypeScript#9841]
**TypeScript Version:**  1.8.0 / nightly (2.0.0-dev.201xxxxx)

**Code**

```ts
interface AllMembersOptional {
    first?: string;
    second?: number;
}

let a: AllMembersOptional = 7;
let b: AllMembersOptional = () => 7;
```

**Expected behavior:**
compiler error '7' is not type AllMembersOptional -- let a: AllMembersOptional = 7;
compiler error 'function: number' is not type AllMembersOptional -- let b: AllMembersOptional = () => 7;

**Actual behavior:** 
no compiler errors or warnings and the following javascript is generated:

```js
var a = 7;
var b = function () { return 7; };
```

**Actual example**
Referencing angular.d.ts and converting an angular directive 

```ts
angular.module('test').directive('custom', function () {
  return {
    controller: 'testCtrl',
    template: '<div></div>,
    //...
  };
});
```
carelessly into a component

```ts
angular.module('test').component('custom', function () {
  return {
    controller: 'testCtrl',
    template: '<div></div>,
    //...
  };
});
```
generates no typescript compiler error messages and will fail at run-time because the options parameter needs to be an object 
```ts
angular.module('test').component('custom', {
    controller: 'testCtrl',
    template: '<div></div>,
    //...
  }
);
```
options parameter cannot a function (or number or string etc). 
Just about anything passed in for a parameter of an interface type with all optional parameters will be allowed by typescript

**Comment by sandersn** [comment]
See [this FAQ answer](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-all-types-assignable-to-empty-interfaces). An interface with all optional properties is similar to an empty interface in this case.

**Comment by bman61** [comment]
It is similar to an empty (i.e. useless) interface but is it the same? 

Per FAQ:
In general, you should _never_ find yourself declaring an interface with no properties.

If _never_ is a good idea, would it make sense to have a compiler flag to issue warnings/errors?

**Comment by sandersn** [comment]
Take a look at #7485 and #3842 -- #7845 is an proposal to fix this based on code from the closed PR #3842.

**Comment by bman61** [comment]
Looked at both. #3842 definitely discusses the issue and issues with implementing. this helps a lot. thanks
